Implement direct pruning

mvandenburgh opened the pull request:
This PR implements the "direct pruning" algorithm described here - https://github.com/spack/spack-infrastructure/blob/0533712e9bf9f9c9075c0aed9d0fc7f720225473/images/ci-prune-buildcache/README.md. 

Currently, the `spack buildcache prune` implements the "orphaned pruning" approach described in the document linked above; this PR extends it to support direct pruning if the user provides a keeplist. Direct pruning is implemented as follows:

- Extract hashes from keeplist
- Iterate over the buildcache, deleting any manifests whose spec hash is not present in the keeplist
- Run existing orphaned pruning
  - This will clean up any blobs associated with manifests that were deleted by the direct pruning

mvandenburgh committed:
Implement direct pruning

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Add tests for direct pruning

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

haampie commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,12 +249,108 @@ def _prune_orphans(
+    """
+    Execute direct pruning for a given mirror using a keeplist file.
+
+    This function reads a file containing package hashes to keep, then deletes
```
you have to be clear what type of hashes you are talking about, package hash means something different in spack.

haampie commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,12 +249,108 @@ def _prune_orphans(
+    Note that this function does *not* prune the blobs associated with the manifests;
+    to do that, `prune_orphan` must be invoked to clean up the now-orphaned blobs.
+
+    :param mirror: The mirror to prune
+    :param keeplist_file: Path to file containing newline-delimited hashes to keep
+    :param dry_run: Whether to perform a dry run without actually deleting
```
wrong style

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/24722d57a6581fd5ae8f724a760f4aaf0821e79b

haampie commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,12 +249,108 @@ def _prune_orphans(
+            line.strip() for line in keeplist_file.read_text().splitlines() if line.strip()
+        )
+    except Exception as e:
+        tty.die(f"Error reading keeplist file {keeplist_file}: {e}")
```
Avoid `tty.die` outside of `spack.cmd`, it prevents spack from being used as a library

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/8125e943218bf07ad27c6bb5dfe46e0f7dcbf4f3

haampie replied:
Hm, you're suppose to `raise SomeSpackError("better message")` or `raise ... from` ;)

mvandenburgh committed:
Replace `tty.die` with `tty.error`

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix docstring style

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix wording

Package hashes are different than spec hashes.

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix edge case (empty buildcache)

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

haampie commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -255,13 +255,13 @@ def prune_direct(mirror: Mirror, keeplist_file: pathlib.Path, dry_run: bool) ->
-    :param keeplist_file: Path to file containing newline-delimited hashes to keep
-    :param dry_run: Whether to perform a dry run without actually deleting
+    Args:
+        mirror (spack.mirrors.mirror.Mirror): Mirror to prune
+        keeplist_file (pathlib.Path): Path to file containing newline-delimited hashes to keep
+        dry_run (bool): Whether to perform a dry run without actually deleting
```
we (or I?) no longer do (type) because it's redundant and may get out of sync with the type hints.

mvandenburgh replied:
Makes sense to me :+1:

mvandenburgh committed:
Remove redundant type info in docstring

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Regenerate shell completions

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Reraise exceptions instead of logging with tty.error

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,17 +250,120 @@ def _prune_orphans(
+        for manifest in manifest_list:
+            cache_entry: Optional[URLBuildcacheEntry] = None
+            try:
+                cache_entry = cast(URLBuildcacheEntry, read_fn(manifest))
```
We can actually get the hash directly from the `manifest` name here without having to fetch anything.

Manifests are named `"{name}-{version}-{hash}"`

so you should be be able to do this

```
spec_name, spec_version, spec_hash = re.match(r"([^ ]+)-([^_ ]+)-([^-_\. ]+)", manifest).groups()
```

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,17 +250,120 @@ def _prune_orphans(
+    try:
+        keep_hashes = set(
+            line.strip() for line in keeplist_file.read_text().splitlines() if line.strip()
+        )
```
Let's also validate that all of the hashes in this list are 32 digits long and error on malformed hashes.

mvandenburgh committed:
Validate keep hashes during pruning

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Remove unnecessary fetching of manifests

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

kwryankrattiger reviewed (dismissed):
I think this looks good for a first pass. We can bike shed other ways to build keep lists that are more spack-like but works like the original build cache pruner.

mvandenburgh commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,17 +252,130 @@ def _prune_orphans(
 
 
-def prune(mirror: Mirror, dry_run: bool) -> None:
+def prune_direct(mirror: Mirror, keeplist_file: pathlib.Path, dry_run: bool) -> None:
```
TODO: add option to keep any blobs uploaded *after* pruning has started

mvandenburgh replied:
@kwryankrattiger do you think there's any value in making this configurable? We originally said yes, but now I'm wondering if it's better to simplify things (both from a maintainer/code perspective and a user/command line interface perspective) and just always ignore new specs uploaded after direct pruning begins.

kwryankrattiger replied:
Oh yeah, I don't think it should be optional. Checking the time stamp should always be required for the direct pruning or it will do odd things.

If someone wants to make it optional later they can, but I don't see a use case for that mode in the near term so probably not worth it.

mvandenburgh committed:
Implement `stat` in web_utils

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Don't delete blobs created after start of pruning

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -408,6 +427,34 @@ def prune_orphan(mirror: Mirror, dry_run: bool) -> None:
+
+        web_util.remove_url(remote_path)
+
+        return start_time
```
This function might fit better in `url_buildcache.py` or something. But because it has a major side effect that isn't apparent without reading the full docstring/code (creating/deleting a file on the buildcache storage), I left it in here so to avoid it becoming a general use function. But I could see reasons to do it either way. @kwryankrattiger any thoughts?

kwryankrattiger replied:
I think it is fine to leave here for now since it is only used by pruning.

mvandenburgh commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -850,3 +852,223 @@ def test_buildcache_prune_orphaned_manifest(tmp_path, mutable_database, mock_gnu
+    # 3. Both manifests should still exist because the race condition logic
+    #    protected them from deletion (they appear to be modified after pruning started)
+    assert web_util.url_exists(manifest_url1)
+    assert web_util.url_exists(manifest_url2)
```
Definitely open to suggestions if there's a cleaner way to test this functionality. I've used libraries like `freezegun` in pytest contexts before to mock the system clock itself, but I went with this moer extensive mocking approach to avoid introducing a new test dependency.

mvandenburgh replied:
And actually, this test has a bug 😬 - will fix

mvandenburgh replied:
Fixed

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -289,3 +429,48 @@ def prune(mirror: Mirror, dry_run: bool) -> None:
+    """
+    with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as f:
+        tmpdir = Path(f)
+        touch_file = tmpdir / ".spack-prune-marker"
```
use a unique name here, uuid or something similar. There is a race when multiple spack instances call this at the same time where the file may be gone before the stat can be called.

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/b4dca7dcd2ab748fd2e90ba163530306f35c1c34

mvandenburgh committed:
Add test for new spec race condition

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Use uuid in prune time function

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Drop usage of `str.removeprefix` (py3.9 only)

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -106,22 +113,29 @@ def _delete_manifests_from_cache_aws(
+        stat_result = web_util.stat_url(url)
+        assert stat_result is not None
+        if stat_result[1] > pruning_started_at:
+            tty.info(f"Skipping deletion of {url} because it was modified after pruning started")
```
This can potentially output a lot, and this case is non-destructive so I don't think reporting it by default is necessarily required.

```suggestion
            tty.verbose(f"Skipping deletion of {url} because it was modified after pruning started")
```

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
+        pruning_started_at: Timestamp of when the pruning started
+        dry_run: Whether to perform a dry run without actually deleting
+    """
+    tty.info("=== Direct Pruning Phase ===")
```
```suggestion
    tty.info("Running Direct Pruning")
```

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
+
+    keep_hashes: Set[str] = set()
+    for line in keeplist_file.read_text().splitlines():
+        keep_hash = line.strip()
```
It may also make sense to support the case where the hash starts with `/` since that is often how Spack formats hashes.

```suggestion
        keep_hash = line.strip()
        if keep_hash[0] == "/":
            keep_hash = keep_hash[1:]
```

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/e98eec8599f0a662494eccee236d0373330aec04

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
+    for line in keeplist_file.read_text().splitlines():
+        keep_hash = line.strip()
+        if len(keep_hash) != 32:
+            raise MalformedKeepListException(f"Found malformed hash in keeplist: {keep_hash}")
```
```suggestion
            raise MalformedKeepListException(f"Found malformed hash in keeplist: {line}")
```

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
+            manifests_to_delete = set(manifests_to_prune)
+
+            # Try AWS CLI deletion first for S3 mirrors
+            if mirror.fetch_url.startswith("s3://"):
```
This pattern is repeated twice now, maybe roll this into a helper function `_delete_manifests_from_cache`

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/ed9284fe7a7678471385ede2b2b8696516632aa0

mvandenburgh replied:
https://github.com/spack/spack/pull/51153/commits/ed9284fe7a7678471385ede2b2b8696516632aa0

kwryankrattiger commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
+
+        if dry_run:
+            for spec_name in specs_to_prune:
+                tty.info(f"  Would prune: {spec_name}")
```
Here we are still printing the specs we will skip due to time boundary. The delete function should just indiscriminately delete everything it is given, the filtering can go in here.

mvandenburgh replied:
The problem with doing the filtering here is that, in the worst case (that being an S3 buildcache with spack running on a system without awscli),  it involves making N http requests to S3 for N files in the S3 bucket (the `stat_url` function). Doing the `stat_url` inside the `_delete_object` allows us to at least parallelize that. 

I see what you're saying about the delete function ideally just deleting things indiscriminately. I can try breaking the `stat_url` call out into its own parallelizable function and see what that looks like.

mvandenburgh replied:
Here is the result of that - https://github.com/spack/spack/pull/51153/commits/c7aa6d1649320b1712adca8b80878c268534b6a7

I think I actually like that approach better. Not having to pass the `pruning_started_at` var through as many functions is nice. wdyt?

kwryankrattiger replied:
I like this approach. The head requests should all be pretty quick to so the overhead vs an `aws ls` is probably acceptable.

mvandenburgh committed:
Improve logging

Co-authored-by: Ryan Krattiger <80296582+kwryankrattiger@users.noreply.github.com>
Signed-off-by: Mike VanDenburgh <37340715+mvandenburgh@users.noreply.github.com>
Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Refactor common code into function

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Refactor mtime comparison logic

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Support `/` prefix in keeplist

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix style

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix type hint

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

kwryankrattiger commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -842,12 +849,21 @@ def migrate_fn(args):
     assert isinstance(mirror, spack.mirrors.mirror.Mirror)
 
-    prune(mirror, dry_run)
+    started_at = get_buildcache_normalized_time(mirror)
```
Thinking more on the right way to get the pruning boundary time. We should try to use the `mtime` of the cache index manifest, then the the normalized time if that fails. This way we at least try to bound the pruning based on mirror state with the assumption that the index is updated as part of the builder workflow (like how we do it is CI).

mvandenburgh committed:
Default to mtime of cache index, if exists

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
strip off parent path before parsing manifest name

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix spec filename regex

The regex didn't account for underscores in the version number before,
for example, this spec:
`openjdk-17.0.11_9-74evnafyw5narzyqa6rc7brfpsovmynt.spec.manifest.json`

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Fix URLs vs file paths handling

In the codepath where AWS CLI is used to interact with a S3 buildcache,
there was some discrepancies where full URL paths (i.e.,
paths including a `file://` or `s3://` prefix) were expected, but
file paths were used (and vice versa).

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

mvandenburgh committed:
Move logic out of cmd module

Moves all nontrivial logic out of the cmd module into the
buildcache_prune module. Aside from being a cleaner design, this will
make it easier to call the pruner from a `spack python` script

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

Signed-off-by: Mike VanDenburgh <michael.vandenburgh@kitware.com>

kwryankrattiger committed:
Merge branch 'develop' into direct-pruning

kwryankrattiger reviewed (approved):
Looks good, testing against public mirrors looks good.
